src/code/audioMgr.c

```diff
@@ -0,0 +1,137 @@
+#include "global.h"
+
+void func_80172C30(AudioMgr* audioMgr) {
+    AudioTask* task = audioMgr->rspTask;
+
+    if (audioMgr->rspTask->taskQueue != NULL) {
+        osSendMesg(task->taskQueue, NULL, OS_MESG_BLOCK);
+    }
+}
+
+void AudioMgr_HandleRetrace(AudioMgr* audioMgr) {
+    static s32 D_801D14F0 = 10;
+    AudioTask* rspTask;
+    s32 timerMsgVal = 666;
+    OSTimer timer;
+    s32 msg;
+
+    if (SREG(20) > 0) {
+        audioMgr->rspTask = NULL;
+    }
+
+    while (audioMgr->unk_F4.validCount != 0) {
+        osRecvMesg(&audioMgr->unk_F4, NULL, OS_MESG_NOBLOCK);
+    }
+
+    if (audioMgr->rspTask != NULL) {
+        audioMgr->audioTask.next = NULL;
+        audioMgr->audioTask.flags = 2;
+        audioMgr->audioTask.framebuffer = NULL;
+
+        audioMgr->audioTask.list = audioMgr->rspTask->task;
+        audioMgr->audioTask.msgQ = &audioMgr->unk_F4;
+
+        audioMgr->audioTask.msg = NULL;
+        osSendMesg(&audioMgr->sched->cmdQ, &audioMgr->audioTask, OS_MESG_BLOCK);
+        Sched_SendEntryMsg(audioMgr->sched);
+    }
+
+    if (SREG(20) >= 2) {
+        rspTask = NULL;
+    } else {
+        rspTask = func_80192BE0();
+    }
+
+    if (audioMgr->rspTask != NULL) {
+        while (true) {
+            osSetTimer(&timer, OS_USEC_TO_CYCLES(32000), 0, &audioMgr->unk_F4, (OSMesg)timerMsgVal);
+            osRecvMesg(&audioMgr->unk_F4, (OSMesg*)&msg, OS_MESG_BLOCK);
+            osStopTimer(&timer);
+            if (msg == timerMsgVal) {
+                Fault_Log("AUDIO SP TIMEOUT %08x %08x\n", audioMgr->rspTask, audioMgr->rspTask->task);
+                if (D_801D14F0 >= 0) {
+                    D_801D14F0--;
+                    func_8017703C(audioMgr->sched);
+                } else {
+                    Fault_Log("audioMgr.c:もうダメ！死ぬ！\n");
+                    osDestroyThread(NULL);
+                    break;
+                }
```

Comment:
This additional code for the timeout handling is very interesting. A little while ago I worked on some `sched.c` functions for mm and found they added some extra stuff there. `func_8017703C` informs the scheduler than it should attempt to "cancel" the current audio task by manually trying to stop the RSP. If it is successful it will re-queue the task and try again. This makes `D_801D14F0` the number of retries and could probably be named `retryCount`.
For anyone wondering, there is also an equivalent for graphics tasks.

Reply:
It's cool finding all the additions/deletions in the MM code relative to OoT's code. Is the convention for using `s` prefix for static variables still used at this low level? I.e. `retryCount` vs `sRetryCount`?

Reply:
I believe we don't use the `s` prefix for in-function statics.